src/phoenix/server/api/routers/auth.py

```diff
@@ -0,0 +1,187 @@
+import asyncio
+from datetime import datetime, timezone
+from functools import partial
+
+from fastapi import APIRouter, Depends, HTTPException, Request, Response
+from sqlalchemy import delete, select
+from sqlalchemy.orm import joinedload
+from starlette.status import HTTP_204_NO_CONTENT, HTTP_401_UNAUTHORIZED, HTTP_404_NOT_FOUND
+
+from phoenix.auth import (
+    PHOENIX_ACCESS_TOKEN_COOKIE_NAME,
+    PHOENIX_ACCESS_TOKEN_MAX_AGE,
+    PHOENIX_REFRESH_TOKEN_COOKIE_NAME,
+    PHOENIX_REFRESH_TOKEN_MAX_AGE,
+    Token,
+    delete_access_token_cookie,
+    delete_refresh_token_cookie,
+    is_valid_password,
+    set_access_token_cookie,
+    set_refresh_token_cookie,
+)
+from phoenix.db.enums import UserRole
+from phoenix.db.models import User as OrmUser
+from phoenix.server.jwt_store import JwtStore
+from phoenix.server.rate_limiters import ServerRateLimiter, fastapi_rate_limiter
+from phoenix.server.types import (
+    AccessTokenAttributes,
+    AccessTokenClaims,
+    AccessTokenId,
+    RefreshTokenAttributes,
+    RefreshTokenClaims,
+    RefreshTokenId,
+    UserId,
+)
+
+rate_limiter = ServerRateLimiter(
+    per_second_rate_limit=0.2,
+    enforcement_window_seconds=30,
+    partition_seconds=60,
+    active_partitions=2,
+)
+login_rate_limiter = fastapi_rate_limiter(rate_limiter, paths=["/login"])
+router = APIRouter(
+    prefix="/auth", include_in_schema=False, dependencies=[Depends(login_rate_limiter)]
+)
+
+
+@router.post("/login")
+async def login(request: Request) -> Response:
+    data = await request.json()
+    email = data.get("email")
+    password = data.get("password")
+
+    if not email or not password:
+        raise HTTPException(status_code=HTTP_401_UNAUTHORIZED, detail="Email and password required")
+
+    async with request.app.state.db() as session:
+        user = await session.scalar(
+            select(OrmUser).where(OrmUser.email == email).options(joinedload(OrmUser.role))
+        )
+        if (
+            user is None
+            or (password_hash := user.password_hash) is None
+            or (salt := user.password_salt) is None
+        ):
+            raise HTTPException(status_code=HTTP_401_UNAUTHORIZED, detail=LOGIN_FAILED_MESSAGE)
+
+    loop = asyncio.get_running_loop()
+    password_is_valid = partial(
+        is_valid_password, password=password, salt=salt, password_hash=password_hash
+    )
+    if not await loop.run_in_executor(None, password_is_valid):
```

Comment:
kinda confused by this. Why do you need to run in the loop here?

Reply:
computing the password hash is by design an expensive operation to mitigate against brute-force attacks. this is an optimization to run the computation in a background thread and yield control of the event loop while it is being computed.